Add/update block height keywords in Clarity

obycode opened the pull request:
### Description

Adding support for `stacks-block-height` and `tenure-height` in Clarity 3. This also includes using the `tenure-height` logic for `block-height` logic in Clarity 1 and 2.

### Applicable issues

- fixes #3943

### Additional info (benefits, drawbacks, caveats)

### Checklist

- [x] Test coverage for new or modified code paths
- [x] Changelog is updated
- [x] Required documentation changes (e.g., `docs/rpc/openapi.yaml` and `rpc-endpoints.md` for v2 endpoints, `event-dispatcher.md` for new events)
- [x] New clarity functions have corresponding PR in `clarity-benchmarking` repo
- [x] New integration test(s) added to `bitcoin-tests.yml`

obycode commented:
I'm still checking on a couple of the integration tests, but @kantai, when you have some time, could you take a look at the approach here and let me know if you see anything troublesome. 🙏

friedger commented:
Somehow related is the get-tenure-info https://github.com/stacks-network/stacks-core/issues/4716
Is that in scope for Clarity 3?

jcnelson commented on clarity/src/vm/contexts.rs:
```diff
@@ -822,6 +822,16 @@ impl<'a, 'hooks> OwnedEnvironment<'a, 'hooks> {
         }
     }
+
+    #[cfg(any(test, feature = "testing"))]
```
Nit: This should go into the test module

obycode replied:
Addressed in 7a63f899e8945a770642517e96324330597dbe87.

obycode replied:
Turns out moving this into the test module creates some difficulty elsewhere because of our use of `#[cfg(any(test, feature = "testing"))]` in various places. I pulled it out of this impl block and put it in its own block with `#[cfg(any(test, feature = "testing"))]` instead.

jcnelson commented on clarity/src/vm/database/clarity_db.rs:
```diff
@@ -855,6 +855,31 @@ impl<'a> ClarityDatabase<'a> {
     }
 
+    fn tenure_height_key() -> &'static str {
+        "_stx-data::tenure_height"
```
Can you make sure SIP-021 reflects this new MARF key?  We've done so for the other MARF keys in the Stacks chainstate MARF.

jcnelson commented on clarity/src/vm/variables.rs:
```diff
@@ -22,31 +24,37 @@ use crate::vm::errors::{InterpreterResult as Result, RuntimeErrorType};
+        NativeVariables::lookup_by_name(name).and_then(|native_variable| {
+            match native_variable.get_max_version() {
+                Some(ref max_version)
+                    if &native_variable.get_min_version() <= version && version <= max_version =>
```
Huh, I didn't know this was legal Rust.  TIL.

jcnelson commented on clarity/src/vm/variables.rs:
```diff
@@ -92,8 +100,18 @@ pub fn lookup_reserved_variable(
+                    let block_height = env.global_context.database.get_current_block_height();
+                    Ok(Some(Value::UInt(block_height as u128)))
+                } else {
+                    let tenure_height = env.global_context.database.get_tenure_height()?;
```
I had to read this a couple times before it made sense.  The `block-height` variable isn't going away _per se_; it's just not recognized as a global in Clarity 3.  But, it still has to have semantic meaning in both Clarity 1 and Clarity 2 contracts while in epoch 3.0.  This code is meant to handle this case specifically -- we evaluate `block-height` as the tenure height when in epoch 3.0.  This comment might need some massaging. 

Related, can you still deploy Clarity 1 and 2 contracts in epoch 3.0?

obycode replied:
Yes, your explanation is correct and is what I was trying to get across with that comment. Does this sound better?

```rust
                // In epoch 2.x, the `block-height` keyword returns the Stacks block height.
                // For Clarity 1 and Clarity 2 contracts executing in epoch 3, `block-height`
                // is equal to the tenure height instead of the Stacks block height. This change
                // is made to maintain a similar pace at which this value increments (e.g. for use
                // as an expiration). In Clarity 3, `block-height` is removed to avoid confusion.
                // It is replaced with two new keywords: `stacks-block-height` and `tenure-height`.
```

> Related, can you still deploy Clarity 1 and 2 contracts in epoch 3.0?
 Yes, you can.

obycode replied:
I updated to the above comment in 7a63f899e8945a770642517e96324330597dbe87.

jcnelson replied:
Thanks for the updated explanation.  But, I'm having a hard time convincing myself that the call to `.get_tenure_height()` can _never_ return an `InterpreterError::Expects(..)` due to the tenure height key not being mapped in the MARF tip of the calling context.  For example, what happens if I do the following?

```clarity
;; Can I call this in Nakamoto?
(at-block SOME_BLOCK_BEFORE_NAKAMOTO tenure-height)`
```

What happens with traits defined in Nakamoto but that get called in a pre-Nakamoto context?

```clarity
;; Suppose .whoops-trait is defined in the Nakamoto epoch, and suppose it has this function
;; definition:
;; (define-public (get-tenure-height) (response uint uint))
(use-trait .whoops-trait)

;; Suppose `tenure-getter`'s `get-tenure-height` function simply returns `tenure-height`.
;; How would this work?
(define-read-only (whoops (tenure-getter <whoops-trait>))
   (at-block SOME_BLOCK_BEFORE_NAKAMOTO (contract-call? tenure-getter get-tenure-height))
```

jcnelson replied:
I have similar questions about `block-height`.  Can these work?

```clarity
;; Can I access `block-height` within an `at-block` closure within Nakamoto as long as the block
;; predates Nakamoto, or is this now forbidden?
(at-block SOME_BLOCK_BEFORE_NAKAMOTO block-height)
```

Again, what about calls to a trait implementation?

```clarity
;; Suppose .whoops-trait is defined in epoch 2.x, and suppose it has this function
;; definition:
;; (define-public (get-block-height) (response uint uint))
(use-trait .whoops-trait)

;; Suppose `block-height-getter`'s `get-block-height` function simply returns `block-height`.
;; How would this work when called in Nakamoto?
(define-public (whoops (block-height-getter <whoops-trait>))
   (at-block SOME_BLOCK_BEFORE_NAKAMOTO (contract-call? block-height-getter get-block-height)))
```

jcnelson commented on clarity/src/vm/variables.rs:
```diff
@@ -123,6 +141,16 @@ pub fn lookup_reserved_variable(
                 let chain_id = env.global_context.chain_id;
                 Ok(Some(Value::UInt(chain_id.into())))
             }
+            NativeVariables::StacksBlockHeight => {
```
These need to be gated

obycode replied:
I believe this is already gated by the `lookup_by_name_at_version` that is the source of the `variable`.

jcnelson commented on stackslib/src/chainstate/stacks/boot/docs.rs:
```diff
@@ -1,4 +1,5 @@
 use clarity::vm::docs::contracts::{produce_docs_refs, ContractSupportDocs};
+use clarity::vm::ClarityVersion;
```
missing copyright comment block

obycode replied:
Added in 7a63f899e8945a770642517e96324330597dbe87.

jcnelson commented on stackslib/src/clarity_vm/clarity.rs:
```diff
@@ -294,6 +295,26 @@ impl ClarityInstance {
             )
         };
 
+        // If we're starting a new tenure, and we're in epoch 3.09, then we
```
Typo -- epoch 3.09?

obycode replied:
Fixed in 7a63f899e8945a770642517e96324330597dbe87.

jcnelson commented on stackslib/src/clarity_vm/clarity.rs:
```diff
@@ -294,6 +295,26 @@ impl ClarityInstance {
 
+        // If we're starting a new tenure, and we're in epoch 3.09, then we
+        // need to increment the tenure height in the Clarity DB.
+        if new_tenure && epoch.epoch_id >= StacksEpochId::Epoch30 {
```
This seems brittle to me.  If you're going to introduce a new MARF key/value pair anyway for representing the tenure height, can you instead update it as part of the Nakamoto chain state's `append_block` function?

obycode replied:
I need the value while building the block, for use while evaluating the transactions included in the block. If we do it in `append_block`, then I think it will be off by 1.

jcnelson replied:
That won't happen?  You can read an inserted a key/value pair into the MARF before committing it, so if you insert the new tenure height before processing Clarity transactions, you should be fine.

obycode replied:
Ah, okay. I was thinking `append_block` was called after the transactions were processed. I'm a little unclear where exactly I would need to put this and where to get the needed information. I can dig into that tomorrow, but if you've got something in mind, the help would be appreciated.

kantai replied:
I think the best strategy is to add this to `setup_block` (which gets invoked in both the miner path and the follower path). Something like this should work:

```diff
diff --git a/stackslib/src/chainstate/nakamoto/mod.rs b/stackslib/src/chainstate/nakamoto/mod.rs
index 2dbcd9f3c..5a9a4b2be 100644
--- a/stackslib/src/chainstate/nakamoto/mod.rs
+++ b/stackslib/src/chainstate/nakamoto/mod.rs
@@ -2551,6 +2551,28 @@ impl NakamotoChainState {
             new_tenure,
         );
 
+        if new_tenure {
+            clarity_tx
+                .connection()
+                .as_free_transaction(|clarity_tx_conn| {
+                    clarity_tx_conn.with_clarity_db(|db| {
+                        let tenure_ht = db.get_tenure_height()?;
+                        db.set_tenure_height(
+                            tenure_ht
+                                .checked_add(1)
+                                .expect("FATAL: tenure height overflowed"),
+                        )?;
+                        Ok(())
+                    })
+                })
+                .map_err(|e| {
+                    error!("Failed to set tenure height during block setup";
+                           "error" => ?e,
+                    );
+                    e
+                })?;
+        }
+
         // now that we have access to the ClarityVM, we can account for reward deductions from
         // PoisonMicroblocks if we have new rewards scheduled
         let matured_rewards_opt = matured_rewards_schedule_opt
```

Then you wouldn't need to pass `new_tenure` through the block begin methods.

obycode replied:
Thanks for this @kantai. Do you think the initialization in `initialize_epoch_3_0` good to stay there?

obycode replied:
Moved in 008f73bf4e5926aa1c8b2c747a06014e37e9e102.

jcnelson replied:
Better yet, you can just call `db.set_tenure_height(coinbase_height)` since it ought to equal to the height of the current tenure.  No need to re-compute a value that's already given!  (cc @kantai for confirmation)

kantai replied:
Yeah -- `coinbase_height` should be the same value.

jcnelson reviewed (commented):
I think there's some more test coverage you'll want to add regarding the interaction between `block-height`, `stacks-block-height`, and `tenure-height`:

* Can you deploy a Clarity 1 or Clarity 2 contract with bound variable names `stacks-block-height` and `tenure-height` in epoch 3.0?

* Can you call into a Clarity 1 and Clarity 2 contract which have bound variable names `stacks-block-height` and `tenure-height` from a Clarity 3 contract, and have those names behave the right way?

* Can a Clarity 1 and Clarity 2 contract take a Clarity 3 trait implementation, whose code body makes use of `stacks-block-height` and `tenure-height`, and have it execute correctly even though the caller is a Clarity 1 or Clarity 2 context?  Can they do so even if they've been deployed before epoch 3?

* Can a Clarity 1 and Clarity 2 contract perform the above using `at-block` to call into the Clarity 3 contract from a block that's before epoch 3.0?

Basically, I'm trying to make sure that these keywords work correctly if control flow passes back and forth across an epoch boundary via either `at-block` or a trait implementation, or a combination of both.

obycode commented:
Thanks for the review @jcnelson. I've addressed the simple comments and will work on adding those tests you requested. I also added an integration test since you reviewed which was the main thing that I was not yet comfortable with.

kantai commented on stacks-common/src/util/macros.rs:
```diff
@@ -88,30 +88,50 @@ macro_rules! define_named_enum {
+            pub fn get_min_version(&self) -> $VerType {
                 match self {
-                    $(
-                        $Name::$Variant => $Version,
-                    )*
+                    $(Self::$Variant => $MinVersion,)*
+                }
+            }
+
+            pub fn get_max_version(&self) -> Option<$VerType> {
+                match self {
+                    $(Self::$Variant => $MaxVersion,)*
```
Can you add rustdocs to the functions e.g., `get_min_version` returns the first Clarity version in which `self` is defined, `get_max_version` returns `Some` for the last Clarity version in which `self` is defined, or `None` if `self` is defined in every Clarity version after `get_min_version()`.

obycode replied:
Added in aa832234791c8d902c44618dabbf0f5d48a7546d.

kantai reviewed (commented):
Looks good to me, just some comments here and there, also responded on the thread about where the best place to perform the `get/set` is.

obycode commented:
Thanks for the review @kantai. Will address these comments shortly.

While adding the tests @jcnelson requested, I realized that I never implemented the analysis error when using `block-height` in Clarity 3. That will be included in the next commit. It's a bit weird, because the rest of the reserved words only hit an error at runtime, but this one will hit the error at analysis. In a future version of Clarity, I'd like to make all of these be detected at analysis time, but not now.

obycode commented on clarity/src/vm/functions/mod.rs:
```diff
@@ -201,10 +201,14 @@ impl NativeFunctions {
-                Some(native_function)
-            } else {
-                None
+            match native_function.get_max_version() {
+                Some(ref max_version)
+                    if &native_function.get_min_version() <= version && version <= max_version =>
+                {
+                    Some(native_function)
+                }
+                None if &native_function.get_min_version() <= version => Some(native_function),
+                _ => None,
```
Sure, I can see how your version is a little easier to read. Will update.

obycode replied:
Refactored into the macro as suggested below.

obycode commented on clarity/src/vm/variables.rs:
```diff
@@ -22,31 +24,37 @@ use crate::vm::errors::{InterpreterResult as Result, RuntimeErrorType};
-                Some(native_function)
-            } else {
-                None
+        NativeVariables::lookup_by_name(name).and_then(|native_variable| {
+            match native_variable.get_max_version() {
+                Some(ref max_version)
+                    if &native_variable.get_min_version() <= version && version <= max_version =>
+                {
+                    Some(native_variable)
+                }
+                None if &native_variable.get_min_version() <= version => Some(native_variable),
+                _ => None,
```
🧠

obycode replied:
Added in aef35a1a57499b0ef3dc111ea0ffd58321eb6875.

obycode commented on clarity/src/vm/types/mod.rs:
```diff
@@ -777,10 +777,14 @@ impl BlockInfoProperty {
+                    Some(native_function)
+                }
+                None if &native_function.get_min_version() <= version => Some(native_function),
+                _ => None,
```
Refactored into the macro as suggested below.

obycode commented:
Ok, I added some more tests in 1bff5b7ce7828248672b74731a8855fa787b06b9. I'm not sure why `test_block_heights_across_versions` is failing yet. I need to resolve that and then add one more test for:

> Can a Clarity 1 and Clarity 2 contract perform the above using at-block to call into the Clarity 3 contract from a block that's before epoch 3.0?

obycode commented on stackslib/src/chainstate/nakamoto/mod.rs:
```diff
@@ -2550,6 +2550,43 @@ impl NakamotoChainState {
+                        let tenure_ht = match db.get_tenure_height() {
+                            Ok(ht) => ht,
+                            Err(e) => {
+                                if cfg!(test) {
+                                    // In test mode, the epoch 3.0 initialization does not always
+                                    // get called to initialize this tenure height so we ignore
+                                    // the error here and return a default value.
+                                    warn!("Failed to get tenure height, defaulting to 0");
+                                    0
+                                } else {
+                                    // If not in test mode, this is definitely an error.
+                                    return Err(e.into());
+                                }
```
I don't love this, but I can't find a better way to handle the test cases that do not properly call the `initialize_epoch_*` functions. Let me know if you have a better solution.

obycode replied:
Probably, but, for example, I was having trouble figuring out how to call it in stackslib/src/chainstate/nakamoto/coordinator/tests.rs, `advance_to_nakamoto`.

kantai replied:
Ah -- I think this was actually a bug.

The `if new_tenure { .. }` block should be _after_ the `process_epoch_transition` invocation.

obycode replied:
🤦‍♂️

obycode replied:
Fixed and cleaned up the tests.

obycode commented:
> Can a Clarity 1 and Clarity 2 contract perform the above using at-block to call into the Clarity 3 contract from a block that's before epoch 3.0?

Is this actually possible? I don't think you can have a `contract-call?` inside an `at-block` because all contract calls are treated as not read-only and only read-only expressions can be inside an `at-block`.

obycode commented:
<img width="819" alt="image" src="https://github.com/stacks-network/stacks-core/assets/1473715/a30b1971-c552-48b5-bbe1-a48b0d159afb">

@wileyj Can we remove the "CI / Rust Format (pull_request_review)" from the required checks?

wileyj commented:
> <img alt="image" width="819" src="https://private-user-images.githubusercontent.com/1473715/329267032-a30b1971-c552-48b5-bbe1-a48b0d159afb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUyNjYyNzgsIm5iZiI6MTcxNTI2NTk3OCwicGF0aCI6Ii8xNDczNzE1LzMyOTI2NzAzMi1hMzBiMTk3MS1jNTUyLTQ4YjUtYmJlMS1hNDhiMGQxNTlhZmIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUwOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MDlUMTQ0NjE4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZGJmYzQ0ZmE3YjJmODczMzFlYjIxYmY1MGRkNDQ0MDI0MzU5ZDRmNjg4MTYyYmExZGViMDhiNTQyNDI4NjM4ZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.HbdS9e8-OekO1WtFj39Kcnd0WROks6iJ48HMmaof8_g">
> @wileyj Can we remove the "CI / Rust Format (pull_request_review)" from the required checks?

it should happen as soon as this is merged: https://github.com/stacks-network/stacks-core/pull/4765

the status checks are valid for any action that triggers the workflow - it's not so fine-grained that we can disable a status check for a specific action (outside of some semi-complicated conditionals for each workflow job).

obycode commented:
I pushed one small change that makes the mutants testing a bit happier. The other items it reports I think should be fixed in a more generic way, or else we'll end up with too many `mutants::skip` annotations throughout the code. E.g. it doesn't detect the tests using rstest macro templates as tests, and so it reports missed mutants for all of those.

jcnelson commented on clarity/src/vm/contexts.rs:
```diff
@@ -1974,6 +1964,18 @@ impl CallStack {
 }
 
+#[cfg(any(test, feature = "testing"))]
+impl<'a, 'hooks> OwnedEnvironment<'a, 'hooks> {
```
Can you put this into the test module that uses it?  Thanks

obycode replied:
I'm not sure who to do that. If I move it into clarity/src/vm/tests/mod.rs, then I cannot access the private field, `self.context`.

jcnelson replied:
You can set `self.contract` to have `pub(crate)` visibility.

jcnelson commented on clarity/src/vm/database/clarity_db.rs:
```diff
@@ -855,13 +856,9 @@ impl<'a> ClarityDatabase<'a> {
     /// Returns the tenure height of the current block.
     pub fn get_tenure_height(&mut self) -> Result<u32> {
-        self.get_data(&Self::tenure_height_key())?
+        self.get_data(TENURE_HEIGHT_KEY)?
```
Should this function and `set_tenure_height` below be gated?

jcnelson replied:
Also, should this fall back to getting the block height if the tenure height key is not mapped in this MARF?  Thinking about what could happen if this function is reachable within a `(at-block SOME_BLOCK_BEFORE_NAKAMOTO ..)` closure.

obycode replied:
Ah, ok, I see what you mean. I was thinking that returning that error would be reasonable if you did something like:

```clarity
(at-block (unwrap-panic (get-block-info? id-header-hash u0)) tenure-height)
```

But maybe your idea of returning the block height would be better.

jcnelson replied:
Ah, I wasn't trying to be prescriptive here.  I was mainly interested in (1) determining if this code could lead to an `InterpreterError::Expects(..)` (or worse), and (2) if so, starting a conversation on what the right behavior ought to be (so we can add regression tests for it).

jcnelson replied:
Also, per your comment earlier, you _can_ have contract-calls within `at-block` as long as they're read-only.  So, it could be possible to call a Clarity 1 or 2 contract from a Clarity 3 function, and it would be possible to do this for either a Nakamoto block or a Stacks epoch2.x block.  I think we should determine what happens if these Clarity 1/2 contracts use `stacks-block-height`, `tenure-height`, and `block-height` when called this way (recalling that Clarity 1/2 contracts can bind these first two global variables as local names, which would now be forbidden).

obycode replied:
Got it. I think it does makes sense to return the block height in that situation. I've got a corresponding test ready for that. I'm a little unsure about whether the gating is necessary for `set_tenure_height`, but I can add it.

jcnelson commented on clarity/src/vm/tests/variables.rs:
```diff
@@ -221,3 +221,466 @@ fn test_tenure_height(
+}
+
+#[apply(test_clarity_versions)]
+fn reuse_block_height(
```
Do you also want to test the following?

* `(define-constant block-height ...)`
* `(define-trait block-height ...)`
* tuples with a key named `block-height`
* `(define-fungible-token ...)`
* `(define-non-fungible-token ...)`

obycode replied:
Sure, I was just trying to cover each category, but I can instead cover all possibilities.

obycode replied:
Added in 890f07cd1d2b939f9f9daa0f497824e7b4359ff7.

jcnelson commented on clarity/src/vm/tests/variables.rs:
```diff
@@ -221,3 +221,466 @@ fn test_tenure_height(
+        "function",
+        r#"
+        (define-private (block-height) true)
+        (define-private (test-func) (block-height))
```
Do you also want to test `(define-public)` and `(define-read-only)`?

obycode replied:
Added in 890f07cd1d2b939f9f9daa0f497824e7b4359ff7.

jcnelson commented on clarity/src/vm/tests/variables.rs:
```diff
@@ -221,3 +221,466 @@ fn test_tenure_height(
+}
+
+#[apply(test_clarity_versions)]
+fn reuse_stacks_block_height(
```
Same concerns here -- there are other Clarity built-ins that let you create named data that need to be tested.

obycode replied:
Added in 890f07cd1d2b939f9f9daa0f497824e7b4359ff7.

jcnelson commented on clarity/src/vm/tests/variables.rs:
```diff
@@ -221,3 +221,466 @@ fn test_tenure_height(
+}
+
+#[apply(test_clarity_versions)]
+fn reuse_tenure_height(
```
Also tagging this for testing with all Clarity built-ins for creating named data.

obycode replied:
Added in 890f07cd1d2b939f9f9daa0f497824e7b4359ff7.

jcnelson commented on stackslib/src/chainstate/nakamoto/mod.rs:
```diff
@@ -2548,9 +2548,29 @@ impl NakamotoChainState {
-            new_tenure,
         );
 
+        if new_tenure {
```
Do you have tests to verify that the tenure height gets initialized to `(+ u1 block-height)` for the first Nakamoto block?

obycode replied:
Are you thinking an integration test for this or some kind of unit test?

obycode replied:
Added to my integration test in 4fb8bdd86cbe108e2f0406ee52cc1d4879b2eb2e.

jcnelson commented on stackslib/src/clarity_vm/tests/contracts.rs:
```diff
@@ -1114,3 +1114,284 @@ fn test_block_heights() {
+            (contract-call? get-trait get-int)
+        )
+        "#;
+    let contract_e3c3 = format!(
```
In a similar vein, what happens if the trait implementation was deployed in epoch 2.x, and had a bound name `stacks-block-height` (or `tenure-height`), and was called in epoch 3.x?  The VM should continue to treat the epoch 2.x's bound names as if they were variables, and not reserved words, but I just want to see a test that confirms this (it may exist; I just may not have gotten to it yet :))

obycode replied:
Added `test_block_heights_across_versions_traits_2_from_3`

jcnelson reviewed (approved):
Is there a test document somewhere that we can update so we can remember to put this PR through its paces on the Nakamoto testnet?

obycode commented:
> Is there a test document somewhere that we can update so we can remember to put this PR through its paces on the Nakamoto testnet?

We probably should start a checklist like https://github.com/stacks-network/stacks-core/discussions/4620. cc @saralab

blockstack-devops commented:
This pull request has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.
